Doc: Dev FAQ Pinned Memory (#3437)
ECP-WarpX/WarpX: Docs/source/developers/faq.rst

This is often seen in a derived class, overwriting an interface method.
It means we do not name the parameter because we do not use it when we overwrite the interface.
But we add the name as a comment ``/* ... */`` so that we know what we ignored when looking at the definition of the overwritten method.

What is Pinned Memory?
----------------------

We need pinned aka "page locked" host memory when we:

> - do async copies from/to host<->device

Reviewer: Thanks for the PR, this is very useful to have. A comment on the English - perhaps this might be more succinct
```suggestion
- do asynchronous copies between the host and device
```

Author: definitely, thanks Dave!

> - want to write to CPU memory from a kernel

Author: ```suggestion
- want to write to CPU memory from a GPU kernel
```

> A typical use case are initialization our (filtered/processed) output routines.

Reviewer 2: ```suggestion
A typical use case is initialization of our (filtered/processed) output routines.
```

AMReX provides pinned memory via the ``amrex::PinnedArenaAllocator`` , which is the last argument passed to constructors of ``ParticleContainer`` and ``MultiFab``.

> Read more on this here: `How to Optimize Data Transfers in CUDA C/C++ <https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/>`__

Author: ```suggestion
Read more on this here: `How to Optimize Data Transfers in CUDA C/C++ <https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/>`__ (note that pinned memory is a host memory feature and works with all GPU vendors we support)
```

Bonus: underneath the hood, asynchronous MPI communications also pin and unpin memory.

> One of the benefits of GPU-aware MPI implementations is, besides possibility to use direct device-device transfers, that MPI and GPU API calls `are aware of each others pinning ambitions <https://www.open-mpi.org/community/lists/users/2012/11/20659.php>`__ and do not create `data races to unpin the same memory <https://github.com/ComputationalRadiationPhysics/picongpu/pull/438>`__.

Reviewer 2: ```suggestion
One of the benefits of GPU-aware MPI implementations is, besides the possibility to use direct device-device transfers, that MPI and GPU API calls `are aware of each others' pinning ambitions <https://www.open-mpi.org/community/lists/users/2012/11/20659.php>`__ and do not create `data races to unpin the same memory <https://github.com/ComputationalRadiationPhysics/picongpu/pull/438>`__.
```

Reviewer 2: I think this is a good idea, esp. the references to more information if a questioner is eager to learn more
